src/components/search/SearchTypesDirectives.js

```diff
@@ -0,0 +1,395 @@
+    'ga_map_service',
+    'ga_marker_overlay_service',
+    'ga_search_service',
+    'ga_urlutils_service',
+    'pascalprecht.translate'
+  ]);
+
+  module.directive('gaSearchLocations',
+      function($http, $q, $sce, $translate, gaUrlUtils, gaBrowserSniffer,
+               gaMarkerOverlay, gaSearchLabels, gaMapUtils, gaDebounce) {
+        return {
+          restrict: 'A',
+          replace: true,
+          templateUrl: 'components/search/partials/searchtypes.html',
+          scope: {
+            options: '=gaSearchLocationsOptions',
+            map: '=gaSearchLocationsMap'
+          },
+          controller: function($scope) {
+            setupController($scope, 'locations', $http, $q, $sce, gaUrlUtils,
+                            gaSearchLabels, gaBrowserSniffer, gaMarkerOverlay);
+          },
+          link: function($scope, element, attrs) {
+            $scope.select = function(res) {
+              unregisterMove();
+              removeOverlay(gaMarkerOverlay, $scope.map);
+              if (originToZoomLevel.hasOwnProperty(res.attrs.origin)) {
+                gaMapUtils.moveTo($scope.map,
+                                  originToZoomLevel[res.attrs.origin],
+                                  [res.attrs.y, res.attrs.x]);
+              } else {
+                gaMapUtils.zoomToExtent($scope.map,
+                                        parseExtent(res.attrs.geom_st_box2d));
+              }
+              addOverlay(gaMarkerOverlay, $scope.map, res);
+              $scope.options.valueSelected(
+                  gaSearchLabels.cleanLabel(res.attrs.label));
+
+              registerMove(gaMarkerOverlay, gaDebounce, $scope.map);
+            };
+
+            $scope.prepareLabel = function(attrs) {
+              var l = gaSearchLabels.highlight(attrs.label,
+                                                 $scope.options.query);
+              if (attrs.origin == 'zipcode') {
+                l = '<span>' + $translate.instant('plz') + ' ' + l +
+                    '</span>';
+              } else if (attrs.origin == 'kantone') {
+                l = '<span>' + $translate.instant('ct') + ' ' + l +
+                    '</span>';
+              } else if (attrs.origin == 'district') {
+                l = '<span>' + $translate.instant('district') + ' ' + l +
+                    '</span>';
+              } else if (attrs.origin == 'parcel') {
+                l += ' <span>' + $translate.instant('parcel') + ' ' +
+                     '</span>';
+              }
+              return $sce.trustAsHtml(l);
+            };
+
+          }
+        };
+      });
+
+  module.directive('gaSearchFeatures',
+      function($rootScope, $http, $q, $sce, $timeout, gaUrlUtils,
+               gaLayerFilters, gaSearchLabels, gaLayers, gaBrowserSniffer,
+               gaMarkerOverlay, gaPreviewFeatures) {
+
+        var selectedFeatures = {};
+        var loadGeometry = function(layerId, featureId, topic, urlbase, cb) {
```

Comment:
Prefer callback here (we don't have cb hell and promise infra requires some overhead).